src/main/java/com/yelp/nrtsearch/server/luceneserver/field/IndexableFieldDef.java

```diff
@@ -110,9 +109,9 @@ protected IndexableFieldDef(String name, Field requestField) {
         }
         childFields.put(childName, (IndexableFieldDef) fieldDef);
       }
-      this.childFields = Collections.unmodifiableMap(childFields);
+      this.childFields = childFields;
     } else {
-      this.childFields = Collections.emptyMap();
+      this.childFields = new HashMap<>();
```

Comment:
This field is intentionally immutable. It is accessed by other indexing threads, and changing it in place can cause issues:
- A ConcurrentModificationException could be thrown in other threads
- The field will be visible to index parsing before it is visible to the ServerCodec, causing the flush to fail
- The field will be visible before the index state is committed, and if there is an error the local state will remain inconsistent

It would be best if the new child field is not visible until the next index state is built. I can't say the best way to do this, without doing some digging. But it may be something like moving the child field maps out of the field classes, and making them accessible through the IndexState.